File: TwoPointers/189.RotateArray.java
// https://leetcode.com/problems/rotate-array/
class Solution {
    public void rotate(int[] nums, int k) {
        int n = nums.length;
        int l = 0;
        int r = n - 1;
        int[] result = new int[n];
        
        for (int i = (k % n) - 1; i >= 0; i--) {
            result[i] = nums[r--];
        }
        
        for (int i = k % n; i < n; i++) {
            result[i] = nums[l++];
        }
        
        for (int i = 0; i < n; i++) {
            nums[i] = result[i];
        }
    }
}

File: TwoPointers/189.RotateArray.swift
// https://leetcode.com/problems/rotate-array/
class Solution {
    func rotate(_ nums: inout [Int], _ k: Int) {
        let n = nums.count
        var l = 0
        var r = n - 1
        
        var result = [Int](repeating: 0, count: n)
        
        var i = (k % n) - 1
        while(i >= 0) {
            result[i] = nums[r]
            r = r - 1
            i = i - 1
        }
        
        var j = k % n
        while (j < n) {
            result[j] = nums[l]
            l = l + 1
            j = j + 1
        }
        
        for (index, num) in result.enumerated() {
            nums[index] = result[index]
        }
    }
}
